subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -529,242 +632,259 @@ func (h *HivedAlgorithm) findVirtualCell(
+		}
+		for podIndex := int32(0); podIndex < int32(len(podPhysicalPlacements)); podIndex++ {
+			mbi.PodPlacements[podIndex].PhysicalGpuIndices = make(
+				[]int32, len(podPhysicalPlacements[podIndex]))
+			mbi.PodPlacements[podIndex].VirtualCellIndices = make(
+				[]int32, len(podPhysicalPlacements[podIndex]))
+			for gpuIndex := 0; gpuIndex < len(podPhysicalPlacements[podIndex]); gpuIndex++ {
+				pGpu := podPhysicalPlacements[podIndex][gpuIndex].(*PhysicalCell)
+				nodes, gpuIndices := pGpu.GetPhysicalPlacement()
+				// here each cell (i.e., pGpu) is only one GPU, hence we takes the first element
+				// in its "nodes" and "gpuIndices" as the node and GPU address
+				if mbi.PodPlacements[podIndex].PhysicalNode == "" {
+					mbi.PodPlacements[podIndex].PhysicalNode = nodes[0]
+				}
+				mbi.PodPlacements[podIndex].PhysicalGpuIndices[gpuIndex] = gpuIndices[0]
+				if groupVirtualPlacement != nil {
+					vGpu := groupVirtualPlacement[podGpuNum][podIndex][gpuIndex].(*VirtualCell)
+					mbi.PodPlacements[podIndex].VirtualCellIndices[gpuIndex] = vGpu.GetIndex()
+				} else {
+					mbi.PodPlacements[podIndex].VirtualCellIndices[gpuIndex] = -1
 				}
 			}
 		}
+		if podGpuNum == currentGpuNum {
+			selectedNode = mbi.PodPlacements[currentPodIndex].PhysicalNode
+			selectedGpuIndices = mbi.PodPlacements[currentPodIndex].PhysicalGpuIndices
+		}
+		affinityGroupBindInfo[groupMemberIndex] = mbi
+		groupMemberIndex++
 	}
-	return cellView
+	return affinityGroupBindInfo, selectedNode, selectedGpuIndices
 }
 
-// buddyAlloc allocates a free cell at a certain level from a cellView.
+// buddyAlloc allocates a free cell at a certain level from a free list.
 // It splits a higher-level cell when there is no free cell at the current level.
-// Note that cellView is one-off: we will calculate it every time we call this function
-// (except recursive calls from it self). So we won't remove a returned cell from cellView.
-func buddyAlloc(cellView ChainCellList, level CellLevel) Cell {
-	if len(cellView[level]) == 0 && level < CellLevel(len(cellView)) {
-		higherCell := buddyAlloc(cellView, level+1)
+// As the input cell list is a copy of the real free list and hence is one-off,
+// we won't remove a returned cell from it.
+func buddyAlloc(freeList ChainCellList, level CellLevel) *PhysicalCell {
+	if len(freeList[level]) == 0 && level < CellLevel(len(freeList)) {
+		higherCell := buddyAlloc(freeList, level+1)
 		if higherCell != nil {
-			cellView[level] = append(cellView[level], higherCell.GetChildren()...)
+			freeList[level] = append(freeList[level], higherCell.GetChildren()...)
 		}
 	}
-	if len(cellView[level]) == 0 {
+	if len(freeList[level]) == 0 {
 		return nil
 	}
-	return minPriorityCell(cellView[level])
-}
-
-func getGroupCellSize(a *api.AffinityGroup) int32 {
-	n := int32(0)
-	for _, m := range a.Members {
-		n += common.NextPowerOf2(m.GpuNumber) * m.PodNumber
+	for _, c := range freeList[level] {
+		fmt.Println(c.GetName())
 	}
```

Comment:
Just for testing?
We should use log instead of Println

Reply:
sorry. forgot to delete it

---
In reply to: [335910134](https://github.com/microsoft/pai/pull/3684#discussion_r335910134) [](ancestors = 335910134)